src/loki-fs-structured-adapter.js

```diff
@@ -0,0 +1,255 @@
+
+/*
+  Loki (node) fs structured Adapter (need to require this script to instance and use it).
+
+  This adapter will save database container and each collection to separate files and
+  save collection only if it is dirty.  It is also designed to use a destructured serialization
+  method intended to lower the memory overhead of json serialization.
+
+  This adapter utilizes ES6 generator/iterator functionality to stream output and
+  uses node linereader module to stream input.  This should lower memory pressure
+  in addition to individual object serializations rather than loki's default deep object
+  serialization.
+*/
+
+(function (root, factory) {
+    if (typeof define === 'function' && define.amd) {
+        // AMD
+        define([], factory);
+    } else if (typeof exports === 'object') {
+        // Node, CommonJS-like
+        module.exports = factory();
+    } else {
+        // Browser globals (root is window)
+        root.LokiFsStructuredAdapter = factory();
+    }
+}(this, function () {
+  return (function() {
+
+    const fs = require('fs');
+    const readline = require('readline');
+    const stream = require('stream');
+
+    /**
+     * Loki structured (node) filesystem adapter class.
+     *     This class fulfills the loki 'reference' abstract adapter interface which can be applied to other storage methods.
+     *
+     * @constructor LokiFsStructuredAdapter
+     *
+     */
+    function LokiFsStructuredAdapter()
+    {
+        this.mode = "reference";
+        this.dbref = null;
+        this.dirtyPartitions = [];
+    }
+
+    /**
+     * Generator for constructing lines for file streaming output of db container or collection.
+     *
+     * @param {object=} options - output format options for use externally to loki
+     * @param {int=} options.partition - can be used to only output an individual collection or db (-1)
+     *
+     * @returns {string|array} A custom, restructured aggregation of independent serializations.
+     * @memberof LokiFsStructuredAdapter
+     */
+    LokiFsStructuredAdapter.prototype.generateDestructured = function*(options) {
+      var idx, sidx;
```

Comment:
as we are in an ES6 context, can we tidy this file up a bit? use `const` for all vars that are not reassigned